### PR TITLE
Feature: Search by user name

### PR DESCRIPTION
**Pull request by lexussuxel**
image
image

### CODE REVIEW

File: packages/frontapp/src/app/components/SearchByUserName/index.tsx [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+    if (search) {
+      navigate(RoutesTypes.READERS);
+      dispatch(setActiveTab(1));
+    }
+  };
+
+  return (
+    <WrapperInput onKeyPress={redirectToSearchByKey}>
+      <StyledSearchIcon
+        fill="#BDBDBD"
```
Comment (unrealsolver) [inline packages/frontapp/src/app/components/SearchByUserName/index.tsx:51]: Please, use constants / ui-kit

File: packages/frontapp/src/app/components/SearchByUserName/index.tsx [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+import { useAppDispatch } from '../../hooks/useTypedDispatch';
+import { useNavigate } from 'react-router-dom';
+import { useGetAllPersonsQuery } from '@mimir/apollo-client';
+import { RoutesTypes } from '../../../utils/routes';
+import { setActiveTab } from '../../store/slices/tabsSlice';
+import { t } from 'i18next';
+import { InputSearch, StyledSearchIcon, WrapperInput } from '../Search';
+import { setSearchReaders } from '../../store/slices/readersSlice';
+
+const SearchByUserName = () => {
```
Comment (unrealsolver) [inline packages/frontapp/src/app/components/SearchByUserName/index.tsx:13]: This looks like a pretty generic search components. It makes sense to refactor it and split into `SearchInput` and `SearchByWhatever`

File: packages/frontapp/src/app/components/SearchByUserName/index.tsx [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+  return (
+    <WrapperInput onKeyPress={redirectToSearchByKey}>
+      <StyledSearchIcon
+        fill="#BDBDBD"
+        width="20"
+        height="20"
+        onClick={redirectToSearchByClick}
+      />
+      <InputSearch
```
Comment (unrealsolver) [inline packages/frontapp/src/app/components/SearchByUserName/index.tsx:56]: Why does this components require so much wrapping? Why `Enter` keypress handling is not default behavior?

File: packages/frontapp/src/app/components/UserList/index.tsx [hunk 4]
```diff
@@ -36,17 +40,23 @@ const Description = styled.p`
...
       <Description>{t('Readers.Description')}</Description>
       <ListWrapper>
-        {data?.getAllPersons.map((person) => (
+        {searchReaders?.map((person) => (
           <SingleUser
-            key={person.id}
-            avatar={person.avatar}
-            name={person.username}
+            avatar={person?.avatar || Pic}
+            name={person?.username || 'a'}
```
Comment (unrealsolver) [inline packages/frontapp/src/app/components/UserList/index.tsx:58]: What does it mean?

**Review by unrealsolver — CHANGES_REQUESTED** [review]

File: packages/apiserver/src/resources/persons/person.resolver.ts [hunk 5]
```diff
@@ -17,12 +17,29 @@ import {
...
+  ) {
+    const data = Person.createQueryBuilder('person')
+      .leftJoinAndSelect('person.location', 'location')
+      .where(
+        `person.location_id= :location ${
+          username ? 'AND person.username ILIKE :name' : ''
+        }`,
+        {
+          location: location_id,
+          name: username ? `%${username}%` : '%',
```
[the remark this answers is not in the document]
Comment (IvanUglovets) [reply packages/apiserver/src/resources/persons/person.resolver.ts:37]: It is bad practice.Please move this logic into a service, the resolver, like the controller, should remain as thin as possible

File: packages/apiserver/src/resources/CurrentUserLocation.decorator.ts [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+      req: { headers },
+    } = ctx.getContext();
+    const bufB64 = Buffer.from(headers['id-token'].split('.')[1], 'base64');
+    const result = JSON.parse(bufB64.toString());
+    const location = await Location.createQueryBuilder('location')
+      .leftJoin('location.person', 'person')
+      .where('person.smg_id= :smg_id', { smg_id: result.sub })
+      .orderBy('location.id', 'DESC')
+      .getOne();
+    return location.id;
```
[the remark this answers is not in the document]
Comment (IvanUglovets) [reply packages/apiserver/src/resources/CurrentUserLocation.decorator.ts:18]: what is this decorator for?

File: packages/frontapp/src/app/components/UserList/index.tsx [hunk 7]
```diff
@@ -36,17 +40,23 @@ const Description = styled.p`
 
 const UserList = () => {
   const { data, loading } = useGetAllPersonsQuery();
+  const dispatch = useAppDispatch();
+  useEffect(() => {
+    if (data) dispatch(setSearchReaders(data?.getAllPersons));
+  }, [data]);
+  const { searchReaders } = useAppSelector((state) => state.readers);
```
Comment (IvanUglovets) [inline packages/frontapp/src/app/components/UserList/index.tsx:47]: it is better to take the state from the redax above before useEffect, so the code looks more beautiful and readable

File: packages/apiserver/src/resources/persons/person.resolver.ts [hunk 8]
```diff
@@ -17,12 +17,29 @@ import {
...
 @Resolver('Person')
 export class PersonResolver {
   @Query(() => [Person])
-  async getAllPersons() {
-    return Person.find();
+  async getAllPersons(
+    @Args('username') username: string,
+    @CurrentUserLocation() location_id: any
+  ) {
+    const data = Person.createQueryBuilder('person')
```
Comment (IvanUglovets) [inline packages/apiserver/src/resources/persons/person.resolver.ts:29]: you forget "await" :)

Comment (IvanUglovets) [reply packages/apiserver/src/resources/persons/person.resolver.ts:29]: also use the try{}catch{} construct because the application might crash(

**Review by IvanUglovets — APPROVED** [review]

**Review by sabrina34352 — APPROVED** [review]

**Pull request merged**